Fix overlapping tiled ljpeg compressed dngs

axxel opened the pull request:
Fix #63 and thereby splitting LJpegPlain in two separate classes dealing with almost completely different use cases. Details see individual commits.

axxel committed:
LJpegDecompressor: Minor code cleanup / preparation of LJpeg DNG fix

* give predictor related variables a more descriptive name
* use std::array for predictor and huffman table array

This is was a 'side' effect of the actual DNG fix (follows next) that I
separated out for better diff-readability.

axxel committed:
Fix overlapping tiled LJpeg compressed DNGs

Border tiles of tiled DNGs may overlap the actual raw image buffer. This
has to be dealt with by decoding but then discarding the pixels that are
overlapping the raw image frame. Fixes #63. It also fixes m/sRaw files
that were recoded into RGB DNGs.

The uncompressed and the compressed-lossless versions of the EOS 5D
files 5G4A9395/5G4A9396 produce the same hash now.

Since this was yet another special casing situation that was affecting
only one of the two use cases (DNG vs CR2), a DNG specific loop has been
added that is a lot easier to read/understand for itself. It turns out,
the DNG and CR2 code paths are almost entirely disjoint -> will separate
them in next commit.

LebedevRI commented on src/librawspeed/decoders/DngDecoderSlices.cpp:
```diff
@@ -418,10 +419,9 @@ void DngDecoderSlices::decodeSlice(DngDecoderThread* t) {
       t->slices.pop();
-      LJpegPlain l(*mFile, e.byteOffset, e.byteCount, mRaw);
-      l.mDNGCompatible = mFixLjpeg;
+      DngDecompressor d(*mFile, e.byteOffset, e.byteCount, mRaw);
```
No.
There is no such thing as 'dng' compression.
It's lossless jpeg compression.
So LJpegDecompressor

LebedevRI replied:
I do see that there already is a LJpegDecompressor cpp/h
It then should be AbstractLJpegDecompressor

axxel replied:
`LJpegDecompressor` is the base class of 3 classes:
  * Cr2Decompressor
  * DngDecompressor
  * HasselbladDecompressor

They are _all_ implementing a form of 'lossless jpeg comression', that is why they share a common base class, but they differ in details specific to the different formats. Hence I think the naming makes absolute sense. You could argue that DNG also has other compression options and calling this one `DngDecompressor` is misleading in the way that it only deals with one variant. So maybe calling it `DngLJpegDecompressor` would be more to your liking?

Introducing the term 'Abstract' would certainly be possible but it is also not used in other places, like `RawDecoder`. Maybe you want to generally introduce this kind naming hint across the library. In that case I'd consider two alternatives, though: either a postfix `Base` or a prefix `I` for interface. Either way: please do that later as a separate pr. To quote yourself: 'one step at a time'.

LebedevRI replied:
> Introducing the term 'Abstract' would certainly be possible <...>. Maybe you want to generally introduce this kind naming hint across the library.

Yeah, it's a flat unstructured mess currently

> but it is also not used in other places, like `RawDecoder`

that is of little concern. the directories weren't used either

axxel replied:
Since the only aspect of my last comment you replied to was the question of whether or not to rename the LJpegDecompressor base class, it is not quite clear to me whether or not you are waiting for me to do that for you. As this has nothing to do with the purpose of this pr or the bug I fixed with it, I will not do that.

LebedevRI replied:
since the last commit already moves stuff around, it kind-of makes sense here.
but no, right now i'd love for travis to come back to life.
so to avoid this naming problem entirely, i'll *try* to take just the fix

axxel replied:
> since the last commit already moves stuff around, it kind-of makes sense here.

Except I did not move/rename _this_ particular class. So it is not like it would need to be renamed again. So merging it as is (maybe rename the DngDecompressor before - see above) and then doing this other improvement later, makes even more sense.

axxel replied:
Now that I decided to put the last patch in a separate pr, it would be easy to change the name of the `DngDecompressor`. Would you please comment on my `DngLJpegDecompressor` suggestion before I push that again?

LebedevRI reviewed (changes requested):
Uhm, @axxel, are you *really* sure the pr fixes the issue?
With the full pr, i get:
```
/Adobe DNG Converter/Canon EOS 5D Mark III/5G4A9394-compressed-lossless.DNG: starting decoding ... 
RawSpeed:Unable to find camera in database: 'Canon' 'Canon EOS 5D Mark III' 'dng'
Please consider providing samples on <https://raw.pixls.us/>, thanks!
=================================================================
==11343==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7f72c09d2780 at pc 0x5645d8695307 bp 0x7f72b57ef3d0 sp 0x7f72b57ef3c8
WRITE of size 2 at 0x7f72c09d2780 thread T8
    #0 0x5645d8695306 in void RawSpeed::DngDecompressor::decodeN<2>()::{lambda(int)#1}::operator()(int) const /home/lebedevri/rawspeed/src/librawspeed/decompressors/DngDecompressor.cpp:94
    #1 0x5645d8695306 in RawSpeed::unroll_loop_t<void RawSpeed::DngDecompressor::decodeN<2>()::{lambda(int)#1}, 1ul>::repeat({lambda(int)#1} const&) /home/lebedevri/rawspeed/src/librawspeed/common/Common.h:224
    #2 0x5645d8695306 in RawSpeed::unroll_loop_t<void RawSpeed::DngDecompressor::decodeN<2>()::{lambda(int)#1}, 2ul>::repeat({lambda(int)#1} const&) /home/lebedevri/rawspeed/src/librawspeed/common/Common.h:223
    #3 0x5645d8695306 in void RawSpeed::unroll_loop<2ul, void RawSpeed::DngDecompressor::decodeN<2>()::{lambda(int)#1}>(void RawSpeed::DngDecompressor::decodeN<2>()::{lambda(int)#1} const&) /home/lebedevri/rawspeed/src/librawspeed/common/Common.h:235
    #4 0x5645d8695306 in void RawSpeed::DngDecompressor::decodeN<2>() /home/lebedevri/rawspeed/src/librawspeed/decompressors/DngDecompressor.cpp:93
    #5 0x5645d869424e in RawSpeed::DngDecompressor::decodeScan() /home/lebedevri/rawspeed/src/librawspeed/decompressors/DngDecompressor.cpp:54
    #6 0x5645d86862c2 in RawSpeed::LJpegDecompressor::parseSOS() /home/lebedevri/rawspeed/src/librawspeed/decompressors/LJpegDecompressor.cpp:127
    #7 0x5645d8688231 in RawSpeed::LJpegDecompressor::decode() /home/lebedevri/rawspeed/src/librawspeed/decompressors/LJpegDecompressor.cpp:46
    #8 0x5645d8694060 in RawSpeed::DngDecompressor::decode(unsigned int, unsigned int, bool) /home/lebedevri/rawspeed/src/librawspeed/decompressors/DngDecompressor.cpp:41
    #9 0x5645d8627c5e in RawSpeed::DngDecoderSlices::decodeSlice(RawSpeed::DngDecoderThread*) /home/lebedevri/rawspeed/src/librawspeed/decoders/DngDecoderSlices.cpp:424
    #10 0x5645d8628fc4 in RawSpeed::DecodeThread(void*) /home/lebedevri/rawspeed/src/librawspeed/decoders/DngDecoderSlices.cpp:65
    #11 0x7f72c641b423 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x7423)
    #12 0x7f72c56849be in __clone (/lib/x86_64-linux-gnu/libc.so.6+0xe89be)

0x7f72c09d2780 is located 0 bytes to the right of 46768000-byte region [0x7f72bdd38800,0x7f72c09d2780)
allocated by thread T0 here:
    #0 0x7f72c6dc6760 in posix_memalign (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc2760)
    #1 0x5645d8661a92 in _aligned_malloc(unsigned long, unsigned long) /home/lebedevri/rawspeed/src/librawspeed/common/Common.cpp:82
    #2 0x5645d85f37b6 in RawSpeed::RawImageData::createData() /home/lebedevri/rawspeed/src/librawspeed/common/RawImage.cpp:92
    #3 0x5645d861dd95 in RawSpeed::DngDecoder::decodeData(RawSpeed::TiffIFD*) /home/lebedevri/rawspeed/src/librawspeed/decoders/DngDecoder.cpp:149
    #4 0x5645d8622ae4 in RawSpeed::DngDecoder::decodeRawInternal() /home/lebedevri/rawspeed/src/librawspeed/decoders/DngDecoder.cpp:300
    #5 0x5645d8649823 in RawSpeed::RawDecoder::decodeRaw() /home/lebedevri/rawspeed/src/librawspeed/decoders/RawDecoder.cpp:304
    #6 0x5645d85ed34e in process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, RawSpeed::CameraMetaData*, bool, bool) /home/lebedevri/rawspeed/src/utilities/rstest/rstest.cpp:207
    #7 0x5645d85eea03 in main._omp_fn.0 /home/lebedevri/rawspeed/src/utilities/rstest/rstest.cpp:295
    #8 0x7f72c5b61d0e in GOMP_parallel (/usr/lib/x86_64-linux-gnu/libgomp.so.1+0xdd0e)
    #9 0x7ffc5fcd23bf  (<unknown module>)

Thread T8 created by T0 here:
    #0 0x7f72c6d34f59 in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x30f59)
    #1 0x5645d8626ecb in RawSpeed::DngDecoderSlices::startDecoding() /home/lebedevri/rawspeed/src/librawspeed/decoders/DngDecoderSlices.cpp:119
    #2 0x5645d861e699 in RawSpeed::DngDecoder::decodeData(RawSpeed::TiffIFD*) /home/lebedevri/rawspeed/src/librawspeed/decoders/DngDecoder.cpp:226
    #3 0x5645d8622ae4 in RawSpeed::DngDecoder::decodeRawInternal() /home/lebedevri/rawspeed/src/librawspeed/decoders/DngDecoder.cpp:300
    #4 0x5645d8649823 in RawSpeed::RawDecoder::decodeRaw() /home/lebedevri/rawspeed/src/librawspeed/decoders/RawDecoder.cpp:304
    #5 0x5645d85ed34e in process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, RawSpeed::CameraMetaData*, bool, bool) /home/lebedevri/rawspeed/src/utilities/rstest/rstest.cpp:207
    #6 0x5645d85eea03 in main._omp_fn.0 /home/lebedevri/rawspeed/src/utilities/rstest/rstest.cpp:295
    #7 0x7f72c5b61d0e in GOMP_parallel (/usr/lib/x86_64-linux-gnu/libgomp.so.1+0xdd0e)
    #8 0x7ffc5fcd23bf  (<unknown module>)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/lebedevri/rawspeed/src/librawspeed/decompressors/DngDecompressor.cpp:94 in void RawSpeed::DngDecompressor::decodeN<2>()::{lambda(int)#1}::operator()(int) const
Shadow bytes around the buggy address:
  0x0feed81324a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feed81324b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feed81324c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feed81324d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feed81324e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0feed81324f0:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feed8132500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feed8132510: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feed8132520: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feed8132530: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feed8132540: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==11343==ABORTING

```

LebedevRI commented:
Also, it would be super cool if we could avoid https://github.com/darktable-org/rawspeed/pull/64#discussion_r98886440 by dropping "Separate LJpegPlain in DngDecompressor and Cr2Decompressor" out of the bugfix pr...

axxel commented:
> Uhm, @axxel, are you really sure the pr fixes the issue?

Outch: no it does not. Embarrassing. Must have slipped through because the different new dngs have not been in my global r.p.u folder as of this morning. The code works but there is an asan issue. Looking into it now.

axxel commented:
Ok. fixed. was one line calculating the width of the right most tile that fits into the raw image frame wrong when N_COMP != mRaw->cpp.

> Also, it would be super cool if we could avoid #64 (comment) by dropping "Separate LJpegPlain in DngDecompressor and Cr2Decompressor" out of the bugfix pr...

I agree that pushing a fix-pr combined with another general improvement pr is not really good style. So I removed it. I hope I got this git-trickery correct.

To my defense for why I added it in the first place: Looking at the fix after it was done, revealed that the structure of the code was simply not fitting the problem properly and it looked 'ugly' to have that "slicesW.size() == 1" special case. And since I spend all the time understanding the differences of the dng vs cr2 use cases I thought it would be a waste of effort if I simply stopped here and let it be.

Now why did I not push it as a separate pr in the first place? Well, you complained about too many prs from my side and did not respond in any way to my concerns as to how I am supposed to interpret that (it makes me feel unwelcome here, in case that was not obvious). I got the message and my fun in putting time in this has dropped significantly. So you'll be likely spared from any major prs from my side for the time being. But since I am also using the code and since I did put all the work in it in the best interest of the project, I'd appreciate it a lot if you would find the time eventually to merge #61, #60 and #54, as well as the upcoming pr with the commit that I just dropped here. It would spare me a lot of effort in maintaining them while still following upstream.

Just to let you know, even though as of now, I feel not welcome anymore, I'll stick to my promise of fixing issues (like this one) should my rewriting/cleanup efforts in the last 2 months be the cause for them.

LebedevRI reviewed (approved):
ok, seems to work.
rstest against 4b452715512f6e70378c941c4240e72b4fd94dc1 does not show any notable breaks.

LebedevRI commented:
> I hope I got this git-trickery correct.

Yes.

> Well, you complained about too many prs from my side and did not respond in any way to my concerns as to how I am supposed to interpret that (it makes me feel unwelcome here, in case that was not obvious).

It is about the amount of changes in general, not just pr. I.e. just squashing everything into just one PR would not workaround the problem.

I have been there, producing much more good important fixes than can be merged. There are pretty much just  two ways here at this moment, as far as i can tell: i could simply ignore all the pr/issues/etc, and occasionally merge something. That is how the prs are currently handled in dt repo.
Or i could have done what i did - inform you that i can not manage to handle all the incoming changes at the same rate you can produce them.
Would the first decision be better for you?
